byteparse/shared/src/main/scala/byteparse/classparse/ClassAttributes.scala

```diff
@@ -0,0 +1,183 @@
+package byteparse
+package classparse
+
+import fastparse.allByte._
+
+object ClassAttributes {
+  import ClassParser.Info._
+  import ClassParser.BasicElems._
+  import ClassParser.Ast._
+  import ClassParser._
+  import CodeParser._
+
+  sealed abstract class Attribute
+
+  case class BasicAttribute(name: String, info: Array[Byte]) extends Attribute {
+    override def equals(other: Any): Boolean =
+      other match {
+        case attr: BasicAttribute => name == attr.name && info.deep == attr.info.deep
+        case _ => false
+      }
+
+    override def toString = {
+      import fastparse.ElemTypeFormatter.ByteFormatter
+      s"BasicAttribute($name,${ByteFormatter.prettyPrint(info)})"
+    }
+  }
+
+  case class ConstantValueAttribute(value: BasicElem) extends Attribute
+
+  case class ExceptionHandler(startIdx: Int, endIdx: Int, handlerIdx: Int, catchType: Option[Class])
+
+  case class CodeAttribute(maxStack: Int, maxLocals: Int,
+                           code: Seq[OpCode], exceptions: Seq[ExceptionHandler],
+                           attributes: Seq[Attribute]) extends Attribute
+
+  case class ExceptionsAttribute(exceptions: Seq[Class]) extends Attribute
+
+
+  case class InnerClass(innerClass: Class, outerClass: Option[Class],
+                        innerName: Option[String], accessFlags: InnerClassFlags)
+
+  case class InnerClassesAttribute(classes: Seq[InnerClass]) extends Attribute
+
+  case class EnclosingMethodAttribute(enclosingClass: Class, enclosingMethodName: Option[String],
+                                      enclosingMethodDescriptor: Option[String]) extends Attribute
+
+  case class SyntheticAttribute() extends Attribute
+
+  case class SignatureAttribute(signature: String) extends Attribute
+
+  case class SourceFileAttribute(sourceFile: String) extends Attribute
+
+  case class DeprecatedAttribute() extends Attribute
+
+  case class BootstrapMethod(bootstrapMethodRef: MethodHandle, bootstrapArguments: Seq[PoolItem])
+
+  case class BootstrapMethodsAttribute(bootstrapMethods: Seq[BootstrapMethod]) extends Attribute
+
+  import ByteUtils.BE._
+
+  val attributeParsers = Map[String, (ClassFileInfo) => Parser[Attribute]](
+    "ConstantValue" -> {
+      (classInfo: ClassFileInfo) =>
+        P( AnyWordI ).map(idx =>
+          ConstantValueAttribute(
+            classInfo.getInfoByIndex[PoolInfo](idx).get match {
+              case BasicElemInfo(elem) => elem
+              case StringInfo(idx) => StringElem(classInfo.getStringByIndex(idx))
+            }))
+    },
+    "Code" -> {
+      (classInfo: ClassFileInfo) => {
+        val exceptionHandler =
```

Comment:
As a convention, could we rewrite these as

```scala
val exceptionHandler = {
  val start_pc = AnyWordI
  val end_pc = AnyWordI
  val handler_pc = AnyWordI
  val catch_type = AnyWordI.map(idx =>
    if (idx == 0) None
    else Some(Class(classInfo, classInfo.getInfoByIndex[ClassInfo](idx).get))
  )
  P( start_pc ~ end_pc ~ handler_pc ~ catch_type).map(ExceptionHandler.tupled)
}
```

I think it would be much easier to read than having the field-names in the comments, and the final `P( start_pc ~ end_pc ~ handler_pc ~ catch_type).map(ExceptionHandler.tupled)` is much easier to skim.

This applies to all the other `P(...)` blocks as well; if it's getting large and gnarly and spread over 8 lines of code, let's extract all the complex bits into local `val`s before and keep the `P(...)` block to 1, maybe 2 lines if possible